qcodes/dataset/data_set_protocol.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import annotations
+
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Dict,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    Sized,
+    Tuple,
+    Union,
+)
+
+import numpy as np
+from typing_extensions import Protocol, runtime_checkable
+
+from qcodes.dataset.descriptions.dependencies import InterDependencies_
+from qcodes.dataset.descriptions.param_spec import ParamSpec, ParamSpecBase
+from qcodes.dataset.descriptions.rundescriber import RunDescriber
+from qcodes.dataset.descriptions.versioning.rundescribertypes import Shapes
+from qcodes.dataset.export_config import DataExportType
+from qcodes.dataset.linked_datasets.links import Link
+
+SPECS = List[ParamSpec]
+# Transition period type: SpecsOrInterDeps. We will allow both as input to
+# the DataSet constructor for a while, then deprecate SPECS and finally remove
+# the ParamSpec class
+SpecsOrInterDeps = Union[SPECS, InterDependencies_]
+
+
+if TYPE_CHECKING:
+    from qcodes.station import Station
+
+    from .data_set_cache import DataSetCache
+
+
+@runtime_checkable
+class DataSetProtocol(Protocol, Sized):
```

Comment:
would it be useful to take the opportunity and group the parts of this interface meaningfully? e.g. group metadata-related things together, properties together, data-related methods together?

however, i can also see that the interface/protocol is huge, so there's no silver-bullter grouping that would generally useful for many readers.

Reply:
If you mean order the members and add some comments in between then yes. We probably cannot split the thing into multiple classes (which I would probably prefer)

Reply:
> If you mean order the members and add some comments in between then yes

yes

Reply:
I tried rearranging them and remove a few members that we could easily remove

Reply:
yes, this looks good now!